api/v1alpha1/cors_types.go

```diff
@@ -26,11 +26,28 @@ import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 // +kubebuilder:validation:Pattern=`^(\*|https?:\/\/(\*|(\*\.)?(([\w-]+\.?)+)?[\w-]+)(:\d{1,5})?)$`
 type Origin string
 
+// RegexOrigin is defined by full-fledged regex, so it is possible to use any
+// scheme, domain, port or any other value. The value should be valid regex
+// string.
+// Syntax should be according to https://github.com/google/re2/wiki/Syntax.
+// Also, backslashes must be escaped.
+//
+// For example, the following are valid origins:
+// - https://foo.example.com
+// - https?://(.*\\.)?example\\.com
+// - .*://localhost(?:\\:\\d+)?
+//
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=253
+type RegexOrigin string
+
 // CORS defines the configuration for Cross-Origin Resource Sharing (CORS).
 type CORS struct {
 	// AllowOrigins defines the origins that are allowed to make requests.
 	// +kubebuilder:validation:MinItems=1
 	AllowOrigins []Origin `json:"allowOrigins,omitempty" yaml:"allowOrigins"`
```

Comment:
Can we just loose the validation on AllowOrigins instead of introuducing an new field?

Reply:
I thought about it, but IMO it could be too radical. As I discovered, Envoy itself uses a lot of [types](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/string.proto#envoy-v3-api-msg-type-matcher-v3-stringmatcher) of matching string. So, my point is, if Envoy's `type.matcher.v3.StringMatcher` `Exact` type is faster than `SafeRegex` so it's really good point to have two options, for fast `Exact` and slower `SafeRegex`

Reply:
I would expected a style as following:
```yaml
AllowOrigins:
  - type: Exact
    exact: xxxxx
  - type: Prefix
    prefrix: yyy
  - type: Regex
    regex: zzzz

```

Reply:
Wildcard is a special case of regex, and it's commonly used for hostname matching.

Reply:
@zirain hm, that's really good idea, but it'll brake back compatibility. Maybe, make this by new field or by issuing new api version

Reply:
I prefer the suggestions provided by @zirain and @zhaohuabing, lets reuse the existing API field
1. the CEL validation will need to be relaxed
2. the logic https://github.com/envoyproxy/gateway/blob/8d82ea9748b57a0dada0bea133fbac018f9ba972/internal/gatewayapi/securitypolicy.go#L523 will need to be revisited to account for `Exact` , `Wildcard Hostnames` and `Regex` based hostnames

Reply:
@arkodg could you clarify your point: do you want to keep the only one `allowOrigins` with zirain's `type` field? Or zhaohuabing's way of two options `allowOrigins` with (Exact and Wildcard) and another with Regex.

Reply:
reuse `AllowOrigins []Origin` for regex as well w/o any breaking change to the API
it looks like contour does the same thing https://projectcontour.io/docs/1.28/config/cors/

Reply:
sure sgtm